File: backend/filearr/agent_about.py
"""Per-agent About / dependency report (``GET /agents/{agent_id}/about``).

WHY this module exists (user request, 2026-08-11): "perform version checking on
dependencies on each agent like the central console in an about page",
"include the tools versions also", "pull the tool locations also".

:mod:`filearr.about` answers "which build is running HERE" by interrogating the
live central process — its Python, its Postgres, its Meilisearch, its host
tools. Until this module existed there was no equivalent question anyone could
ask an AGENT. The console knew one string about a remote machine's software,
``agent_version``, and that is a release stamp: it says nothing about the Go
toolchain that compiled the binary, the modules linked into it, the host OS, or
which copy of exiftool the service will actually execute.

**THE GOVERNING DIFFERENCE FROM about.py: nothing here is probed.** about.py
reads live numbers because it is running on the machine it is describing. This
module is describing a machine on the other side of a network that central
cannot — and by design must not — reach out to. So every fact here is an agent
SELF-REPORT, arriving on the 60-second command poll and stored verbatim in
``agents.capabilities``, and the honest framing of every value is "this is what
the agent said, as of ``capabilities_at``". The endpoint is pure assembly: no
network, no subprocess, one row read.

**ONE COMPARATOR, still.** The agent reports facts — version strings, paths —
and central alone judges them, through the same :mod:`filearr.toolversions`
registry and :mod:`filearr.versioncmp` comparator that judge central's own tools
on the About page. Nothing in the Go agent compares a version to a minimum, and
nothing in the frontend does either. A poppler called outdated on one page and
fine on another would discredit both at once, and the fastest way to arrive
there is a second implementation of "is 24.02.0 newer than 22.09.0".

**NEVER BLANK FOR UNKNOWN.** Same rule as about.py and about.ts. An agent that
has never advertised capabilities returns ``reported: False`` with empty
sections — not an error, and emphatically not zeros or seven ``absent`` chips,
which would state as observed fact something never observed. A tool present but
silent about its version is ``version: None`` with verdict ``unknown``, which
the console renders as "installed, version unknown".

**Everything in ``capabilities`` is third-party JSON** from a machine central
does not control, so every field is type-checked on the way out rather than
trusted. A malformed or hostile advertisement degrades to a missing section; it
never raises.
"""

from __future__ import annotations

from typing import Any

from filearr import agentsync, toolversions
from filearr.models import Agent

#: Cap on any single string copied out of an agent's advertisement. These are
#: untrusted strings from a remote machine that end up rendered in a browser and
#: pasted into bug reports; a version banner or a path has no length contract on
#: the wire. Matches the spirit of ``about._MAX_TEXT``.
_MAX_TEXT = 400

#: Cap on the module list central will echo back to the console. The agent
#: already caps itself at 200 (``maxModules`` in
#: ``agent/internal/inventory/modules.go``); this is the belt to that braces,
#: because the cap that matters is the one on the side that does not trust the
#: input.
_MAX_MODULES = 300

#: The fields of the agent's ``build`` block that are rendered, in the order the
#: console shows them. An allowlist rather than a passthrough: this dict comes
#: from a remote machine and a future agent adding a key must not be able to
#: inject an unbounded blob into every console page load.
_BUILD_FIELDS = (
    "go_version",
    "goos",
    "goarch",
    "os_version",
    "vcs_revision",
    "vcs_time",
    "vcs_modified",
    "main_version",
    "num_cpu",
)


def _text(value: Any) -> str | None:
    """A capped, whitespace-collapsed string, or ``None`` for anything that is
    not a usable string. Non-strings become ``None`` rather than ``str(value)``:
    rendering ``"None"`` or ``"{}"`` in a table cell is worse than an honest gap.
    """
    if not isinstance(value, str):
        return None
    flat = " ".join(value.split())
    return flat[:_MAX_TEXT] or None


def _dict(value: Any) -> dict:
    return value if isinstance(value, dict) else {}


def build_section(capabilities: dict | None) -> dict | None:
    """The agent's build provenance, or ``None`` when it reported none.

    ``None`` (rather than a dict of nulls) is what lets the console say "this
    agent's build is not reported — it predates the About channel, or it has not
    polled since" instead of drawing an empty table that looks like a fault.

    ``go_version`` is the toolchain that COMPILED the binary, not a pin in
    go.mod, and the console labels it "built with" for exactly the reason
    about.py separates a pyproject pin from an installed distribution's
    metadata: intent and fact are different questions and they disagree
    precisely when it matters.
    """
    raw = _dict(_dict(capabilities).get("build"))
    if not raw:
        return None
    out: dict[str, Any] = {}
    for key in _BUILD_FIELDS:
        value = raw.get(key)
        if value is None:
            continue
        if isinstance(value, bool):
            out[key] = value
        elif isinstance(value, int):
            out[key] = value
        elif (text := _text(value)) is not None:
            out[key] = text
    # Present the allowlisted keys uniformly: a field the agent did not send is
    # explicitly null, so the console renders "not reported" rather than having
    # to distinguish a missing key from a null one.
    return {key: out.get(key) for key in _BUILD_FIELDS} if out else None


def host_tool_rows(capabilities: dict | None) -> list[dict]:
    """One row per host tool the agent knows about, judged by central.

    Row shape mirrors :func:`filearr.about.host_tools` EXACTLY —
    ``{name, purpose, present, version, path, url, minimum_version, verdict,
    impact}`` — so the console's two tables are the same table with a different
    subject, and one component's cell logic (``about.ts`` / ``agentAbout.ts``)
    reads both.

    Order is :data:`filearr.toolversions.HOST_TOOLS`, then any tool the agent
    advertises that this central release has never heard of, sorted. Including
    the unknown ones is deliberate: a newer agent build that grows an eighth
    tool must show up here rather than vanish, and it lands on ``unknown``
    (no published minimum), which is exactly the right verdict for a tool we
    have no opinion about.

    ``path`` is the resolved absolute location on the agent host. It answers
    "which copy", which "present: true" cannot, and it is also the visible proof
    of an agent-side security rule: the agent resolves host tools only from
    machine-wide, admin-writable directories, because it runs as
    LocalSystem/root and executing a binary out of a user-writable path would be
    a local privilege-escalation. A path here under a user profile means that
    rule regressed.
    """
    caps = _dict(capabilities)
    tools = _dict(caps.get("tools"))
    versions = _dict(caps.get("tool_versions"))
    paths = _dict(caps.get("tool_paths"))

    names = list(toolversions.HOST_TOOLS) + sorted(
        n for n in tools if isinstance(n, str) and n not in toolversions.HOST_TOOLS
    )
    rows = []
    for name in names:
        present = tools.get(name) is True
        version = _text(versions.get(name))
        rows.append(
            {
                "name": name,
                "purpose": toolversions.tool_purpose(name),
                "present": present,
                "version": version,
                # Only a present tool can have a location. An agent that sent a
                # path for a tool it also reported absent is contradicting
                # itself; ``present`` is the field the capability matrix already
                # trusts, so it wins.
                "path": _text(paths.get(name)) if present else None,
                "url": toolversions.tool_url(name),
                "minimum_version": toolversions.minimum_version(name),
                "verdict": toolversions.tool_verdict(name, present, version),
                "impact": toolversions.minimum_impact(name),
            }
        )
    return rows


def module_rows(capabilities: dict | None) -> list[dict] | None:
    """The Go modules linked into the agent binary, or ``None`` if not reported.

    ``None`` and ``[]`` mean different things and the console renders them
    differently: ``None`` is "the agent did not send a module list" (an older
    build, or the payload budget — see ``modules_omitted``), while ``[]`` would
    claim the binary has no dependencies. Only the first is ever produced here.

    ``url`` is derived, never looked up: ``https://pkg.go.dev/<path>`` is
    deterministic for every public module, and this page — like the central
    About page — makes no outbound network calls. A private module path simply
    yields a link that 404s at pkg.go.dev, which is a better failure than a
    page that hangs on a box with no route to the internet.
    """
    raw = _dict(capabilities).get("modules")
    if not isinstance(raw, list):
        return None
    rows = []
    for entry in raw[:_MAX_MODULES]:
        if not isinstance(entry, dict):
            continue
        path = _text(entry.get("path"))
        if not path:
            continue
        rows.append(
            {
                "path": path,
                "version": _text(entry.get("version")),
                "url": f"https://pkg.go.dev/{path}",
            }
        )
    return rows or None


def extract_section(capabilities: dict | None) -> dict:
    """What this agent's extraction pass can do — the compact restatement of the
    capability advertisement that the About panel shows next to the tools.

    Kept separate from the chips row on the Agents page rather than replacing
    it: the chips are the at-a-glance answer scanned across a fleet, this is the
    detailed one read about a single machine.
    """
    caps = _dict(capabilities)
    formats = caps.get("formats")
    collectors = caps.get("inventory_collectors")
    schema = caps.get("extract_schema")
    return {
        "schema": schema if isinstance(schema, int) else None,
        "formats": [f for f in formats if isinstance(f, str)] if isinstance(formats, list) else [],
        "collectors": (
            [c for c in collectors if isinstance(c, str)] if isinstance(collectors, list) else []
        ),
        "inventory_version": (
            caps.get("inventory_version")
            if isinstance(caps.get("inventory_version"), int)
            else None
        ),
    }


def _int(value: Any) -> int | None:
    """A non-negative int, or ``None``. ``bool`` is rejected explicitly because
    it is an ``int`` subclass in Python and ``True`` rendering as ``1`` in a
    counter column would be an invented fact. Negatives are rejected for the same
    reason: every value here is a count or a byte size."""
    if isinstance(value, bool) or not isinstance(value, int) or value < 0:
        return None
    return value


def rehash_section(health: dict | None) -> dict | None:
    """The agent's quick_hash migration state (QH-T6), or ``None`` when it has
    never run a sweep.

    **Why this reads HEALTH and not CAPABILITIES**, unlike everything else in
    this module: capabilities is a slow-moving statement about what an agent CAN
    do, while this changes minute by minute during a multi-hour sweep, and health
    is the block the agent re-evaluates on every poll.

    **Why the console needs it at all.** This is the ONLY place the migration's
    state is visible, and central cannot derive it from the catalogue. There is
    no hash-provenance column for agent-owned rows — ``agentsync.apply_batch``
    never writes ``policy_version`` — so no query central can write distinguishes
    a stale agent quick_hash from a correct one. The agent's own cursor is the
    entire completion signal, and this is the wire it arrives on.

    ``None`` (rather than a dict of zeros) is what lets the console say "not run"
    instead of drawing a converged-looking report of a sweep that never happened.
    Same rule as :func:`build_section`, and the same reason.

    Everything here is third-party JSON from a machine central does not control,
    so every field is type-checked on the way out; a malformed block degrades to
    ``None`` rather than raising.
    """
    raw = _dict(_dict(health).get("rehash"))
    started = _text(raw.get("started"))
    if not started:
        # The agent omits the whole block until a sweep has run, and stamps
        # started_at before its first batch — so no start time means no sweep,
        # whatever else the block happens to contain.
        return None
    return {
        # "h<scheme>-<min>-<max>", deliberately human-readable on the Go side so
        # an operator can compare two agents' migration state by eye.
        "fp": _text(raw.get("fp")),
        "started": started,
        "finished": _text(raw.get("finished")),
        # The agent's own flag rather than ``finished is not None``: an empty
        # finished_at is the durable "partway through" state the next command
        # resumes from, and it is the agent's to declare.
        "complete": raw.get("complete") is True,
        "seen": _int(raw.get("seen")),
        # changed vs verified stay SEPARATE all the way to the panel. A sweep
        # reporting 40,000 seen / 0 changed / 40,000 verified is confirming an
        # already-converged agent; the same numbers with the two added together
        # are indistinguishable from a sweep that did nothing.
        "changed": _int(raw.get("changed")),
        "verified": _int(raw.get("verified")),
        "skipped": _int(raw.get("skipped")),
        "failed": _int(raw.get("failed")),
        "cursor": _int(raw.get("cursor")),
        "min_size": _int(raw.get("min_size")),
        "max_size": _int(raw.get("max_size")),
    }


def agent_about(agent: Agent) -> dict[str, Any]:
    """Assemble the whole per-agent About report from one already-loaded row.

    Pure and synchronous — no session, no I/O, no network — so it is trivially
    testable against a hand-built ``capabilities`` dict, which is the point:
    every interesting case here is a shape a remote machine sent us, and the
    only way to have confidence in the degradation paths is to be able to write
    them down.

    ``reported`` is the flag the console keys its empty state on. False means
    this agent has never polled with a capability advertisement — a pending
    enrollment, an agent that has not started yet, or a build older than the
    capability channel. That is a NORMAL state, not an error, and the endpoint
    returns 200 with empty sections rather than a 404 or a page of zeros.
    """
    caps = agent.capabilities if isinstance(agent.capabilities, dict) else None
    return {
        "agent": {
            "id": str(agent.id),
            "name": agent.name,
            "hostname": agent.hostname,
            "platform": agent.platform,
            "status": agentsync.agent_status(agent),
            # The release stamp the agent binary carries. Deliberately shown
            # next to, never instead of, the build block: this is what the
            # pipeline named the release, while `build.vcs_revision` is what the
            # binary was actually made from, and a hand-built or re-tagged
            # binary is exactly the case where they disagree.
            "agent_version": agent.agent_version,
            "config_generation_applied": agent.config_generation_applied,
            "last_seen_at": agent.last_seen_at,
            # Central's OBSERVATION of the transport ('bearer' | 'mtls'), not
            # the agent's claim about it.
            "auth_mode": agent.last_auth_mode,
            # When the advertisement below arrived. Everything in this report is
            # as-of this instant, and on an offline agent that can be days ago —
            # which is why it is a first-class field rather than a footnote.
            "capabilities_at": agent.capabilities_at,
        },
        "build": build_section(caps),
        "host_tools": host_tool_rows(caps) if caps else [],
        "modules": module_rows(caps),
        # The agent sets this when its own dependency list would have pushed the
        # poll payload past its self-imposed budget. Distinguishing "did not
        # send" from "deliberately left out, and here is why" is the difference
        # between a console that says "not reported (payload budget)" and one
        # that shows a confident empty table.
        "modules_omitted": _dict(caps).get("modules_omitted") is True,
        "extract": extract_section(caps),
        # QH-T6. Sourced from HEALTH, not capabilities (see rehash_section), and
        # so it is present even on an agent that has never sent a capability
        # advertisement — hence its own null rather than a field of `extract`.
        "rehash": rehash_section(
            agent.health if isinstance(agent.health, dict) else None
        ),
        "reported": caps is not None,
    }

File: frontend/src/lib/agentAbout.ts
// Per-agent About panel — the DOM-free half.
//
// Sibling of ./about.ts, and deliberately written in its voice: the row shapes
// the API returns, the "how do I render an unknown value honestly" decisions,
// and the Markdown dump behind the Copy button. Kept out of the .svelte
// component so it can be unit-tested on Node without a bundler or a DOM (see
// frontend/tests/agentAbout.node.test.ts).
//
// THE ONE RULE, inherited from ./about.ts: an unknown value never renders as a
// blank or a zero. "not reported", "version unknown", "not installed" are all
// legitimate, meaningful answers, and collapsing any of them to an empty cell
// turns "we could not determine this" into "there is nothing here".
//
// THE SECOND RULE, inherited from ./hostTools.ts: this module does not judge
// versions. Central computed every `verdict` (backend/filearr/toolversions.py),
// and the status chips in the panel come from `toolChip` in ./hostTools — there
// is exactly one comparator in this codebase and it is in Python, because the
// About page's server-probed tools and an agent's advertised tools must be
// judged by the same rule or neither page is believed.
//
// THE THIRD RULE, specific to this surface: every fact here is a REMOTE
// SELF-REPORT, not a live probe. Central never reaches out to an agent — agents
// poll — so the whole panel is "what this agent last said, at
// `capabilities_at`". On an offline agent that can be days old and still look
// perfectly current, which is why the timestamp is rendered at the top rather
// than tucked into a tooltip.

// Explicit `.ts` extension: this is a VALUE import (the two formatters are
// reused rather than reimplemented — a second `shortSha` that truncated to a
// different length would make the two pages disagree about the same commit),
// and the Node test runner resolves it at runtime with no bundler to guess an
// extension for it. Every other cross-module import in src/lib/ is type-only
// and therefore erased before it reaches Node.
import { formatWhen, shortSha, type Cell } from "./about.ts";

export type AgentAboutIdentity = {
  id: string;
  name: string;
  hostname: string;
  platform: string;
  status: string;
  agent_version: string | null;
  config_generation_applied: number | null;
  last_seen_at: string | null;
  /** Central's OBSERVATION of the transport ('bearer' | 'mtls'), not a claim
   *  the agent made about itself. */
  auth_mode: string | null;
  /** When the advertisement this whole panel is derived from arrived. */
  capabilities_at: string | null;
};

export type AgentAboutBuild = {
  /** The Go toolchain that COMPILED the binary — not a pin in go.mod. Labelled
   *  "built with" in the UI for that reason. */
  go_version: string | null;
  goos: string | null;
  goarch: string | null;
  os_version: string | null;
  vcs_revision: string | null;
  vcs_time: string | null;
  /** Working tree was dirty at build time: the commit id alone does not
   *  identify this binary. */
  vcs_modified: boolean | null;
  main_version: string | null;
  num_cpu: number | null;
};

/** One host tool on the AGENT's machine. Structurally the same row as
 *  `AboutTool` (./about.ts) — same keys, same meanings — because the two tables
 *  answer the same question about different machines. */
export type AgentAboutTool = {
  name: string;
  purpose: string | null;
  present: boolean;
  version: string | null;
  /** Resolved absolute path ON THE AGENT HOST. */
  path: string | null;
  url: string | null;
  minimum_version: string | null;
  verdict: "ok" | "outdated" | "unknown" | "absent";
  impact: string | null;
};

export type AgentAboutModule = {
  path: string;
  version: string | null;
  url: string;
};

export type AgentAboutExtract = {
  schema: number | null;
  formats: string[];
  collectors: string[];
  inventory_version: number | null;
};

export type AgentAbout = {
  agent: AgentAboutIdentity;
  build: AgentAboutBuild | null;
  host_tools: AgentAboutTool[];
  /** Null = no module list was reported. Never `[]`, which would claim the
   *  binary has no dependencies. */
  modules: AgentAboutModule[] | null;
  /** True when the AGENT deliberately left its module list out of the poll to
   *  stay inside its payload budget. */
  modules_omitted: boolean;
  extract: AgentAboutExtract;
  /** False = this agent has never advertised capabilities at all. */
  reported: boolean;
};

/** Render any value that might be missing. The fallback is a SENTENCE, never an
 *  empty string — see the module note. */
export function orUnknown(v: string | number | null | undefined, fallback = "not reported"): string {
  if (v === null || v === undefined) return fallback;
  const s = String(v).trim();
  return s === "" ? fallback : s;
}

/** How a host-tool row reads on this panel.
 *
 *  Identical decisions to `toolCell` in ./about.ts — the states and their
 *  wording must match, or the same poppler reads differently depending on which
 *  page you are on — with one addition this surface needs: the resolved PATH is
 *  a first-class column here rather than tooltip-only, because "which copy of
 *  exiftool will the service run" is a question you can only ask about a remote
 *  machine you are not sitting at.
 *
 *  The verdict is central's (`t.verdict`), never recomputed. */
export function agentToolCell(t: AgentAboutTool): Cell {
  if (!t.present) return { text: "not installed", tone: "muted" };
  if (!t.version)
    return {
      text: "installed, version unknown",
      tone: "muted",
      hint:
        "The binary resolved but would not state a version (an older build, or a " +
        "wrapper script), so it cannot be checked against a recommended minimum.",
    };
  if (t.verdict === "outdated")
    return {
      text: `${t.version} — below ${t.minimum_version ?? "the recommended minimum"}`,
      tone: "warn",
      hint:
        `Filearr recommends ${t.minimum_version ?? "a newer version"} or newer.` +
        (t.impact ? ` ${t.impact}` : "") +
        " Fixing it is a host-side package upgrade on the agent machine — no new agent build.",
    };
  if (t.verdict === "unknown")
    // Installed, states a version, and we still decline to judge: either the
    // string carries no comparable number (a git build, usually NEWER than any
    // release) or Filearr publishes no minimum. Neutral on purpose.
    return {
      text: t.version,
      tone: "ok",
      hint: t.minimum_version
        ? `This version cannot be compared with the recommended minimum of ${t.minimum_version}, so it is not judged either way.`
        : "Filearr publishes no minimum version for this tool.",
    };
  return {
    text: t.version,
    tone: "ok",
    hint: t.minimum_version ? `Minimum recommended: ${t.minimum_version}.` : undefined,
  };
}

/** How the LOCATION cell reads.
 *
 *  Absent tools have no location and say so. A present tool with no path is a
 *  contradiction the agent should never send, but if it does the honest answer
 *  is that we were not told — not a blank cell that reads as "installed at
 *  nowhere".
 *
 *  `bad` tone is reserved for a path under a user profile. That must never
 *  happen: the agent resolves host tools only from machine-wide,
 *  admin-writable directories, because it runs as LocalSystem/root and
 *  executing a binary out of a user-writable directory would let any local user
 *  escalate. Displaying the path is what makes that rule auditable from the
 *  console, and this is the alarm if it ever regresses. */
export function toolPathCell(t: AgentAboutTool): Cell {
  if (!t.present) return { text: "—", tone: "muted" };
  if (!t.path) return { text: "location not reported", tone: "muted" };
  if (isUserProfilePath(t.path))
    return {
      text: t.path,
      tone: "bad",
      hint:
        "This tool resolved from a USER-WRITABLE directory. The agent is supposed to " +
        "refuse those: it runs as LocalSystem/root, so executing a binary out of a " +
        "user profile is a local privilege-escalation path. Report this — it means " +
        "the agent's resolver regressed.",
    };
  return { text: t.path, tone: "ok" };
}

/** Does a path live under a user profile? Deliberately conservative and
 *  case-insensitive — it only ever adds a warning, never removes one. */
export function isUserProfilePath(path: string): boolean {
  const p = path.replace(/\//g, "\\").toLowerCase();
  return (
    p.includes("\\users\\") ||
    p.startsWith("c:\\users") ||
    /^\\home\\[^\\]+\\/.test(p) ||
    /^\\users\\[^\\]+\\/.test(p)
  );
}

/** The build stack as label/value rows, in the order the panel renders them.
 *
 *  `go_version` is labelled "Go toolchain (built with)" on purpose: it is the
 *  compiler, not the `go` directive in go.mod, and the two disagree routinely.
 *  Central's About page draws the same distinction between a pyproject pin
 *  (intent) and installed metadata (fact). */
export function buildRows(a: AgentAbout): { label: string; value: string; hint?: string }[] {
  const b = a.build;
  const rows: { label: string; value: string; hint?: string }[] = [
    {
      label: "Agent version",
      value: orUnknown(a.agent.agent_version),
      hint: "The release stamp the agent binary carries — what the pipeline named it.",
    },
  ];
  rows.push({
    label: "Go toolchain",
    value: orUnknown(b?.go_version),
    hint: "The Go toolchain that COMPILED this binary — not the `go` directive in go.mod.",
  });
  rows.push({
    label: "Built for",
    value:
      b?.goos && b?.goarch ? `${b.goos}/${b.goarch}` : orUnknown(b?.goos ?? b?.goarch ?? null),
  });
  rows.push({ label: "Host OS", value: orUnknown(b?.os_version) });
  rows.push({
    label: "CPUs",
    value: orUnknown(b?.num_cpu),
    hint: "Logical CPUs visible to the agent process.",
  });
  rows.push({
    label: "Source commit",
    value: vcsText(b),
    hint: b?.vcs_modified
      ? "The working tree was DIRTY at build time, so this commit does not fully identify the binary."
      : b?.vcs_time
        ? `Committed ${formatWhen(b.vcs_time)}.`
        : undefined,
  });
  rows.push({
    label: "Reported",
    value: a.agent.capabilities_at ? formatWhen(a.agent.capabilities_at) : "never",
    hint:
      "When this agent last sent its capability advertisement. Everything on this " +
      "panel is as of that moment — central never queries an agent, agents poll.",
  });
  return rows;
}

function vcsText(b: AgentAboutBuild | null | undefined): string {
  const sha = shortSha(b?.vcs_revision);
  if (!sha) return "not reported";
  return b?.vcs_modified ? `${sha} (modified)` : sha;
}

/** What the modules section should say. Three states, and conflating the last
 *  two is the mistake this function exists to prevent: "the agent did not tell
 *  us" and "there are none" look identical as an empty table. */
export function modulesSummary(a: AgentAbout): Cell {
  if (a.modules && a.modules.length)
    return { text: `${a.modules.length} module(s) linked into the agent binary`, tone: "ok" };
  if (a.modules_omitted)
    return {
      text: "not reported (payload budget)",
      tone: "muted",
      hint:
        "The agent left its dependency list out of the poll to stay inside the " +
        "capability payload budget. The agent's own local web UI shows it in full — " +
        "nothing crosses a network to reach that page.",
    };
  return {
    text: "not reported",
    tone: "muted",
    hint: "This agent build does not report its Go module list.",
  };
}

function mdTable(headers: string[], rows: string[][]): string {
  return [
    `| ${headers.join(" | ")} |`,
    `| ${headers.map(() => "---").join(" | ")} |`,
    ...rows.map((r) => `| ${r.join(" | ")} |`),
  ].join("\n");
}

/** Escape the two characters that break a Markdown table cell. Versions and
 *  paths are third-party strings and a Windows path is full of backslashes. */
function md(value: string | number | null | undefined, fallback = "unknown"): string {
  const v = String(value ?? "").trim();
  if (!v) return fallback;
  return v.replace(/\|/g, "\\|").replace(/\n/g, " ");
}

function link(text: string, url: string | null | undefined): string {
  return url ? `[${md(text)}](${url})` : md(text);
}

/**
 * The whole agent report as Markdown, for pasting into a bug report — the
 * realistic reason anyone opens this panel and then reaches for the mouse.
 *
 * The "as of" line is not decoration: a pasted report of a machine's tool
 * versions is worthless without the timestamp, and an agent that has been
 * offline for a week produces a perfectly confident-looking dump.
 *
 * `now` is injectable so the output is deterministic under test.
 */
export function agentAboutMarkdown(a: AgentAbout, now = new Date()): string {
  const parts: string[] = [];
  const id = a.agent;

  parts.push(`# Agent ${md(id.name)} — build stack`);
  parts.push(
    `_Captured ${now.toISOString()} from central. The agent last reported at ` +
      `${id.capabilities_at ? md(id.capabilities_at) : "never"} — every value below is as of then._`,
  );

  if (!a.reported) {
    parts.push(
      "**This agent has never sent a capability advertisement.** It has not polled " +
        "yet (a fresh enrollment), or it is running a build older than the capability " +
        "channel. Nothing below is missing — there is nothing yet to report.",
    );
  }

  parts.push(
    mdTable(
      ["Field", "Value"],
      [
        ["Agent", md(id.name)],
        ["Hostname", md(id.hostname)],
        ["Platform", md(id.platform)],
        ["Status", md(id.status)],
        ["Transport", md(id.auth_mode, "not observed")],
        ["Last seen", md(id.last_seen_at, "never")],
        ...buildRows(a).map((r) => [md(r.label), md(r.value)]),
      ],
    ),
  );

  parts.push("## Host tools (on the agent machine)");
  parts.push(
    a.host_tools.length === 0
      ? "Not reported."
      : mdTable(
          ["Tool", "Status", "Path"],
          a.host_tools.map((t) => [
            link(t.name, t.url),
            md(agentToolCell(t).text),
            md(t.path, "—"),
          ]),
        ),
  );

  parts.push("## Extraction");
  parts.push(
    mdTable(
      ["Field", "Value"],
      [
        ["Schema", md(a.extract.schema)],
        ["Formats", md(a.extract.formats.join(", "), "none")],
        ["Inventory collectors", md(a.extract.collectors.join(", "), "none")],
        ["Inventory version", md(a.extract.inventory_version)],
      ],
    ),
  );

  parts.push("## Go modules");
  parts.push(
    a.modules && a.modules.length
      ? mdTable(
          ["Module", "Version"],
          a.modules.map((m) => [md(m.path), md(m.version, "unknown")]),
        )
      : md(modulesSummary(a).text),
  );

  return parts.join("\n\n") + "\n";
}
